File: Week_01/id_33/1array/threeSum/threeSum.go
package main

import (
	"fmt"
	"sort"
)

/**
排序后，先取a值，通过从两边逼近去b,c
*/
func threeSum(nums []int) [][]int {
	var results [][]int
	n := len(nums)
	if n < 3 {
		return results
	}
	// 排序后防止重复
	sort.Ints(nums)
	for i := 0; i < n-2; i++ {
		a := nums[i]
		if i > 0 && nums[i-1] == a {
			continue
		}
		left := i + 1
		right := n - 1
		for left < right {
			b := nums[left]
			c := nums[right]
			if a+b+c == 0 {
				results = append(results, []int{a, b, c})
				left++
				right--
				for left < right && nums[left] == nums[left-1] {
					left++
				}
				for left < right && nums[right] == nums[right+1] {
					right--
				}
			} else if a+b+c < 0 {
				left++
			} else {
				right--
			}
		}
	}
	return results
}

func main() {
	//nums := []int{-1, 0, 1, 2, -1, -4}
	nums := []int{0, 0, 0, 0}
	fmt.Println(threeSum(nums))
}

File: Week_01/id_5/threeSum.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */

var threeSum = function (nums) {
  let res = []
  let size = nums.length
  nums.sort((a, b) => a - b)
  if (nums[0] > 0 || nums[size - 1] < 0) return []
  for (let i = 0; i < size - 2;) {
    let left = i + 1
    let right = size - 1
    do {
      if (nums[i] * nums[right] > 0) break
      let sum = nums[i] + nums[left] + nums[right]
      if (sum === 0) {
        res.push([nums[i], nums[left], nums[right]])
      }
      if (sum <= 0) {
        while (nums[left] === nums[++left]) {}
      } else {
        while (nums[right] === nums[--right]) {}
      }
    } while (left < right)
    while (nums[i] === nums[++i]) {}
  }
  return res
}